library/psa_crypto.c

```diff
@@ -4881,26 +4832,21 @@ static psa_status_t psa_generate_derived_key_internal(
 
 #if defined(MBEDTLS_PSA_BUILTIN_KEY_TYPE_ECC_KEY_PAIR) || \
     defined(MBEDTLS_PSA_BUILTIN_KEY_TYPE_ECC_PUBLIC_KEY) || \
-    defined(MBEDTLS_PSA_ACCEL_KEY_TYPE_ECC_KEY_PAIR) || \
-    defined(MBEDTLS_PSA_ACCEL_KEY_TYPE_ECC_PUBLIC_KEY)
+    defined(MBEDTLS_PSA_BUILTIN_ALG_ECDSA) || \
+    defined(MBEDTLS_PSA_BUILTIN_ALG_DETERMINISTIC_ECDSA) || \
+    defined(MBEDTLS_PSA_BUILTIN_ALG_ECDH)
     if ( PSA_KEY_TYPE_IS_ECC( slot->attr.type ) )
     {
         if ( PSA_KEY_TYPE_ECC_GET_FAMILY( slot->attr.type ) != PSA_ECC_FAMILY_MONTGOMERY )
```

Comment:
Please avoid `if (condition expressed negatively) ... else ...`. Generally, please avoid double negations: it gets hard to keep track of them (was it not equal, or not not equal, or not not not equal?).

One of these days we'll add Edwards curves in addition to Montgomery and Weierstrass. So I think we'll need to introduce a macro `PSA_ECC_FAMILY_IS_WEIERSTRASS` (and maybe `IS_MONTGOMERY` as well, but this one is likely to remain forever implemented as just `== PSA_ECC_FAMILY_MONTGOMERY`). Now would be a good time to introduce this macro.

The implementation of the macro can use the structure of family encodings. For the foreseeable future, a correct implementation would be `(family & 0xc0) == 0`.

Reply:
I can add macro `PSA_ECC_FAMILY_IS_WEIERSTRASS`.

I see only the following curve families in code:

```
#define PSA_ECC_FAMILY_SECP_K1           ((psa_ecc_family_t) 0x17)
#define PSA_ECC_FAMILY_SECP_R1           ((psa_ecc_family_t) 0x12)
#define PSA_ECC_FAMILY_SECP_R2           ((psa_ecc_family_t) 0x1b)
#define PSA_ECC_FAMILY_SECT_K1           ((psa_ecc_family_t) 0x27)
#define PSA_ECC_FAMILY_SECT_R1           ((psa_ecc_family_t) 0x22)
#define PSA_ECC_FAMILY_SECT_R2           ((psa_ecc_family_t) 0x2b)
#define PSA_ECC_FAMILY_BRAINPOOL_P_R1    ((psa_ecc_family_t) 0x30)
#define PSA_ECC_FAMILY_MONTGOMERY        ((psa_ecc_family_t) 0x41)
#define PSA_ECC_FAMILY_TWISTED_EDWARDS   ((psa_ecc_family_t) 0x42)
```

There is no WEIERSTRASS family defined. So how do we know that the following condition can be used to determine if curve is Weierstrass `(family & 0xc0) == 0` ?

Reply:
All of these families are Weierstrass curves except the last two (Montgomery and Edwards). Historical note: initially all standardized curves were Weierstrass, so that's not reflected in the family names, because it was assumed. Montgomery and Edwards are more recent additions to the world of ECC.